arlas-core/src/main/java/io/arlas/server/utils/ParamsParser.java

```diff
@@ -266,128 +278,143 @@ public static Filter getFilter(List<String> filters, List<String> q, List<String
-        filter.notpwithin = getValidGeoFilters(notpwithin, true);
-        filter.notgwithin = getValidGeoFilters(notgwithin);
-        filter.notgintersect = getValidGeoFilters(notgintersect);
         filter.dateformat = dateFormat;
         return filter;
     }
 
-    public static List<MultiValueFilter<String>> getValidGeoFilters(List<String> geoStrings) throws ArlasException {
-        return getValidGeoFilters(geoStrings, false);
+    private static boolean isPwithin(ElasticAdmin elasticAdmin, CollectionReference collectionReference, String field, String op) throws ArlasException {
+        if (GEO_OP_WITHIN.contains(OperatorEnum.valueOf(op))) {
+            return CollectionCache.getInstance().getType(elasticAdmin, collectionReference, field) ==  ElasticType.GEO_POINT;
+        }
+        return false;
+    }
+
+    public static Filter getFilterWithValidGeos(ElasticAdmin elasticAdmin, CollectionReference collectionReference, Filter filter) throws ArlasException {
+        Filter newFilter = new Filter();
+        newFilter.q = filter.q;
+        newFilter.dateformat = filter.dateformat;
+        if (filter.f != null) {
+            newFilter.f = new ArrayList<>();
+            for (MultiValueFilter<Expression> orFiltersList : filter.f) {
+                MultiValueFilter<Expression> newOrFiltersList = new MultiValueFilter<>();
+                for (Expression orCond : orFiltersList) {
+                    newOrFiltersList.add(getValidGeoFilter(elasticAdmin, collectionReference, orCond));
+                }
+                newFilter.f.add(newOrFiltersList);
+            }
+        }
+        return newFilter;
+    }
+
+    public static Expression getValidGeoFilter(ElasticAdmin elasticAdmin, CollectionReference collectionReference, Expression expression) throws ArlasException {
+        if (GEO_OP.contains(expression.op)) {
+            expression.value = getValidGeoFilters(expression.value,
+                    isPwithin(elasticAdmin, collectionReference, expression.field, expression.op.name()));
+        }
+        return expression;
     }
 
-    public static List<MultiValueFilter<String>> getValidGeoFilters(List<String> geoStrings, boolean isPwithin) throws ArlasException{
-        if (geoStrings != null) {
-            List<MultiValueFilter<String>> geoFilters =  getStringMultiFilter(geoStrings);
-            List<MultiValueFilter<String>> validGeoFilters = new ArrayList<>();
-            if(geoFilters != null) {
-                for (MultiValueFilter<String> geoFilter : geoFilters) {
-                    MultiValueFilter<String> validGeoFilter = new MultiValueFilter<>();
-                    for (String geo : geoFilter) {
-                        if (CheckParams.isBboxMatch(geo)) {
-                            CheckParams.checkBbox(geo);
-                            validGeoFilter.add(geo);
+    public static String getValidGeoFilters(String geo, boolean isPwithin) throws ArlasException{
```

Comment:
maybe we can rename this method to `getValidGeometry` as it transforms the geometring string only

Reply:
Fixed